homeassistant/components/epson/media_player.py

```diff
@@ -82,34 +88,31 @@
 )
 
 
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
-    """Set up the Epson media player platform."""
-    if DATA_EPSON not in hass.data:
-        hass.data[DATA_EPSON] = []
-
-    name = config.get(CONF_NAME)
-    host = config.get(CONF_HOST)
-    port = config.get(CONF_PORT)
-    ssl = config[CONF_SSL]
-
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Epson projector from a config entry."""
+    timeout_scale = config_entry.options.get(TIMEOUT_SCALE, 1.0)
     epson_proj = EpsonProjector(
-        async_get_clientsession(hass, verify_ssl=False), name, host, port, ssl
+        async_get_clientsession(hass, verify_ssl=config_entry.data[CONF_SSL]),
+        config_entry.title,
+        config_entry.data[CONF_HOST],
+        config_entry.data[CONF_PORT],
+        timeout_scale,
+        config_entry.entry_id,
     )
-
-    hass.data[DATA_EPSON].append(epson_proj)
-    async_add_entities([epson_proj], update_before_add=True)
+    hass.data[DOMAIN][config_entry.entry_id][DATA_EPSON].append(epson_proj)
```

Comment:
Why are you storing entities in `hass.data` ?

Reply:
this way I could retrieve EpsonProjector(MediaPlayerEntity) in `async_service_handler` and call `select_cmode`.
Should I drop this function from `EpsonProjector(MediaPlayerEntity)` and call function of the lib directly `self._projector.send_command(CMODE_LIST_SET[cmode])`? 
I can access `Projector` object of the lib pretty easy now

Reply:
The service handler should be done via [`platform.async_register_entity_service`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/sonos/media_player.py#L392-L400). Since that calls functions on entities, you can just refer to the projector inside the entity.